didcomm/pack.py

```diff
@@ -0,0 +1,128 @@
+from didcomm.did_doc.did_resolver import DIDResolver
+from didcomm.plaintext import Plaintext, PlaintextOptionalHeaders
+from didcomm.secrets.secrets_resolver import SecretsResolver
+
+
+@dataclass(frozen=True)
+class PackResult:
+    """
+    Result of pack operation.
+
+    Attributes:
+        packed_msg (str): a packed message as a JSON string ready to be forwarded to the returned 'service_endpoint'
+        service_endpoint (str): an optional service endpoint to be used to transport the 'packed_msg'.
+    """
+    packed_msg: JSON
+    service_endpoint: Optional[str]
+
+
+@dataclass(frozen=True)
+class PackConfig:
+    """
+    Pack configuration.
+
+    Default config performs repudiable authentication encryption (auth_crypt)
+    and prepares a message ready to be forwarded to the returned endpoint (via Forward protocol).
+
+    Attributes:
+        secrets_resolver (SecretsResolver): an optional secrets resolver that can override a default secrets resolver
+        registered by 'register_default_secrets_resolver'
+
+        did_resolver (DIDResolver): an optional DID Doc resolver that can override a default DID Doc resolver
+        registered by 'register_default_did_resolver'
+
+        enc_alg_auth (AuthCryptAlg): the encryption algorithm to be used for authentication encryption (auth_crypt).
+        `A256CBC_HS512_ECDH_1PU_A256KW` by default.
+
+        enc_alg_anon (AnonCryptAlg): the encryption algorithm to be used for anonymous encryption (anon_crypt).
+        `XC20P_ECDH_ES_A256KW` by default.
+
+        encryption (bool): whether the plaintext needs to be encrypted. True by default.
+        authentication (bool): whether the plaintext needs to be authenticated. True by default.
+        anonymous_sender (bool): whether the sender ID needs to be protected. False by default.
+
+        forward (bool):  whether the packed messages need to be wrapped into Forward messages to be sent to Mediators
+        as defined by the Forward protocol. True by default.
+    """
+    secrets_resolver: SecretsResolver = None
+    did_resolver: DIDResolver = None
+    enc_alg_auth: AuthCryptAlg = AuthCryptAlg.A256CBC_HS512_ECDH_1PU_A256KW
+    enc_alg_anon: AnonCryptAlg = AnonCryptAlg.XC20P_ECDH_ES_A256KW
+    encryption: bool = True
+    authentication: bool = True
+    anonymous_sender: bool = False
+    forward: bool = True
+
+
+@dataclass(frozen=True)
+class PackParameters:
+    """
+    Optional paraneters for pack.
+
+    Attributes:
+        forward_headers (PlaintextOptionalHeaders): if forward is enabled (true by default),
+        optional headers can be passed to the wrapping Forward messages.
+
+        sign_frm (DID_OR_DID_URL): if non-repudiation is needed, a DID or key ID to be used for signing must be specified.
+        Not required by default as repudiation is expected in most of the cases.
+    """
+    forward_headers: Optional[PlaintextOptionalHeaders] = None
+    sign_frm: Optional[DID_OR_DID_URL] = None
```

Comment:
Will this error if a consumer chooses auth_crypt and forgets this, or chooese anon_crypt and enters it? Would it be clearer to split into separate `PackParameter` objects or builders to avoid this?

Reply:
`sign_frm` is an optional parameter for signature (non-repudiation) which can be added in addition to auth_crypt or anon_crypt.
So, if the user doesn't specify it, just auth_crypt (or anon_crypt) will be produced. If the user specified `sign_frm`, then auth_crypt(signed) or anon_crypt(signed) will be produced. 
It seems that DID Comm spec suggests to use authcrypt and keep repudiation by default (see for example [#242](https://github.com/decentralized-identity/didcomm-messaging/issues/242)), that's why by default signatures are not added and there is no need to specify `sign_frm`.

As for sender DID or key to be used for encryption, this is part of `pack` and a required argument ('frm'). Again, in most cases the user will call auth_crypt, so `frm` is a required parameter there.